pir/pir-internal/src/main/java/com/duckduckgo/pir/internal/common/PirActionsRunner.kt

```diff
@@ -0,0 +1,944 @@
+        if (url == DBP_INITIAL_URL) {
+            nextCommand(
+                HandleBroker(
+                    commandsFlow.value.state.copy(
+                        currentBrokerIndex = 0,
+                        currentActionIndex = 0,
+                        profileQuery = profileQuery,
+                    ),
+                ),
+            )
+        } else if (commandsFlow.value is LoadUrl) {
+            // If the current action is still navigate, it means we just finished loading and we can proceed to next action.
+            // Sometimes the loaded url gets redirected to another url (could be different domain too) so we can't really check here.
+            logcat { "PIR-RUNNER (${this@RealPirActionsRunner}): Completed loading for ${commandsFlow.value}" }
+            nextCommand(
+                ExecuteBrokerAction(
+                    commandsFlow.value.state.copy(
+                        currentActionIndex = commandsFlow.value.state.currentActionIndex + 1,
+                    ),
+                ),
+            )
+        } else {
+            logcat { "PIR-RUNNER (${this@RealPirActionsRunner}): Ignoring $url as next action has been pushed" }
+        }
+    }
+
+    private suspend fun handleCommand(
+        command: Command,
+        continuationResult: Continuation<Result<Unit>>,
+    ) {
+        logcat { "PIR-RUNNER ($this): Handle command: $command" }
+        when (command) {
+            Idle -> {} // Do nothing
+            is HandleBroker -> handleBrokerAction(command.state)
+            is ExecuteBrokerAction -> executeBrokerAction(command.state, command.actionRequestData)
+            is CompleteExecution -> {
+                continuationResult.resume(Result.success(Unit))
+                cleanUpRunner()
+            }
+            // TODO add loading timeout
+            is LoadUrl -> coroutineScope.launch(dispatcherProvider.main()) {
+                detachedWebView!!.loadUrl(command.urlToLoad)
+            }
+
+            is BrokerCompleted -> handleBrokerCompleted(command.state, command.isSuccess)
+            is GetEmail -> handleGetEmail(command.state)
+            is AwaitEmailConfirmation -> handleEmailConfirmation(command.state, command.pollingIntervalSeconds)
+            is GetCaptchaSolution -> handleGetCaptchaSolution(command.state, command.responseData, command.isRetry)
+            is AwaitCaptchaSolution -> handleAwaitCaptchaSolution(command.state, command.pollingIntervalSeconds, command.retries, command.attempt)
+            is SendCaptchaSolution -> handleSendCaptchaSolution(command.state, command.callback)
+        }
+    }
+
+    private fun nextCommand(command: Command) {
+        coroutineScope.launch {
+            commandsFlow.emit(command)
+        }
+    }
+
+    private suspend fun handleSendCaptchaSolution(
+        state: State,
+        callback: String,
+    ) {
+        withContext(dispatcherProvider.main()) {
```

Comment:
Seems like this doesn’t show anything visually but i can see the actions passing.